FreeRTOS-Plus/Demo/AWS/Device_Defender_Windows_Simulator/Device_Defender_Demo/metrics_collector.c

```diff
@@ -0,0 +1,243 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * http://www.FreeRTOS.org
+ * http://aws.amazon.com/freertos
+ *
+ * 1 tab == 4 spaces!
+ */
+
+/* Standard includes. */
+#include <stdio.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+
+/* Metrics includes. */
+#include "tcp_netstat.h"
+
+/* Demo config. */
+#include "demo_config.h"
+
+/* Interface include. */
+#include "metrics_collector.h"
+
+/**
+ * @brief The maximum length of line read from any of /proc/net/dev, /proc/net/tcp
+ * and /proc/net/udp.
+ */
+#define MAX_LINE_LENGTH                  ( 256 )
+
+/**
+ * @brief Various connection eStatus.
+ */
+#define CONNECTION_STATUS_LISTEN         ( 10 )
+#define CONNECTION_STATUS_ESTABLISHED    ( 1 )
+
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t xGetNetworkStats( NetworkStats_t * pxOutNetworkStats )
+{
+    MetricsCollectorStatus_t eStatus = MetricsCollectorSuccess;
+
+    MetricsType_t xMetrics = { 0 };
+    BaseType_t xMetricsStatus = 0;
+
+    if( pxOutNetworkStats == NULL )
+    {
+        LogError( ( "Invalid parameter. pxOutNetworkStats NULL" ) );
+        eStatus = MetricsCollectorBadParameter;
+    }
+
+    if( eStatus == MetricsCollectorSuccess )
+    {
+        /* Start with everything as zero. */
+        memset( pxOutNetworkStats, 0, sizeof( NetworkStats_t ) );
+
+        xMetricsStatus = vGetMetrics( &xMetrics );
+
+        if( xMetricsStatus != 0 )
+        {
+            eStatus = MetricsCollectorCollectionFailed;
+        }
+    }
+
+    if( eStatus == MetricsCollectorSuccess )
+    {
+        pxOutNetworkStats->ulBytesReceived = xMetrics.xInput.uxByteCount;
+        pxOutNetworkStats->ulPacketsReceived = xMetrics.xInput.uxPacketCount;
+        pxOutNetworkStats->ulBytesSent = xMetrics.XOutput.uxByteCount;
+        pxOutNetworkStats->ulPacketsSent = xMetrics.XOutput.uxPacketCount;
+    }
+
+    return eStatus;
+}
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t xGetOpenTcpPorts( uint16_t * pusOutTcpPortsArray,
+                                           uint32_t ulTcpPortsArrayLength,
+                                           uint32_t * pulOutNumTcpOpenPorts )
+{
+    MetricsCollectorStatus_t status = MetricsCollectorSuccess;
+
+    MetricsType_t xMetrics = { 0 };
+    BaseType_t xMetricsStatus = 0;
+
+    if( pulOutNumTcpOpenPorts == NULL )
+    {
+        LogError( ( "Invalid parameter. pulOutNumTcpOpenPorts NULL." ) );
+        status = MetricsCollectorBadParameter;
+    }
+
+    if( status == MetricsCollectorSuccess )
+    {
+        xMetricsStatus = vGetMetrics( &xMetrics );
```

Comment:
This is a placeholder function?

Reply:
That function is from the FreeRTOS+TCP tcp_netstat utility